src/image/BackupIncrements.cc

```diff
@@ -0,0 +1,193 @@
+/* -------------------------------------------------------------------------- */
+/* Copyright 2002-2023, OpenNebula Project, OpenNebula Systems                */
+/*                                                                            */
+/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
+/* not use this file except in compliance with the License. You may obtain    */
+/* a copy of the License at                                                   */
+/*                                                                            */
+/* http://www.apache.org/licenses/LICENSE-2.0                                 */
+/*                                                                            */
+/* Unless required by applicable law or agreed to in writing, software        */
+/* distributed under the License is distributed on an "AS IS" BASIS,          */
+/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
+/* See the License for the specific language governing permissions and        */
+/* limitations under the License.                                             */
+/* -------------------------------------------------------------------------- */
+
+#include "BackupIncrements.h"
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int BackupIncrements::from_xml_node(const xmlNodePtr node)
+{
+    int rc = _template.from_xml_node(node);
+
+    if (rc == -1)
+    {
+        return -1;
+    }
+
+    increments.init(&_template);
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+
+int BackupIncrements::add_increment(const std::string& source, long long size,
+        Increment::Type type)
+{
+    VectorAttribute * va = increments.new_increment(source, size, type);
+
+    if (va == nullptr)
+    {
+        return -1;
+    }
+
+    _template.set(va);
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+
+int BackupIncrements::last_increment_id()
+{
+    Increment * li = increments.last_increment();
+
+    if ( li == nullptr )
+    {
+        return -1;
+    }
+
+    return li->id();
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+VectorAttribute * IncrementSet::new_increment(const std::string& source, long long sz,
+        Increment::Type type)
+{
+    Increment * li = last_increment();
+
+    int parent = -1;
+
+    if ( li == nullptr )
+    {
+        if ( type != Increment::FULL )
+        {
+            return nullptr;
+        }
+    }
+    else
+    {
+        parent = li->id();
+    }
+
+    int iid = parent + 1;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

We observed that an integer overflow arising for your program. It happens when the outcome of an arithmetic operation surpasses the maximum representable value for the utilized data type. It can lead to unpredictable behavior, security vulnerabilities, and system crashes. To address this vulnerability, it is essential to perform thorough range checking and validation of input values. Employ data types with adequate size to accommodate the anticipated range of values and implement checks to identify and manage potential overflow conditions.